adal/src/telemetry/java/com/microsoft/aad/adal/APIEvent.java

```diff
@@ -0,0 +1,172 @@
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package com.microsoft.aad.adal;
+
+import android.content.Context;
+import android.util.Pair;
+
+import java.io.UnsupportedEncodingException;
+import java.net.URL;
+import java.security.NoSuchAlgorithmException;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * This class tracks flows for certain API calls. Most notably in those all acquireToken* calls
+ * All other Events will be called within the timeline of APIEvent
+ */
+class APIEvent extends DefaultEvent {
+
+    private static final String TAG = DefaultEvent.class.getSimpleName();
+    private final String mEventName;
+
+    APIEvent(final String eventName) {
+        setEvent(EventStrings.EVENT_NAME, eventName);
+        mEventName = eventName;
+    }
+
+    APIEvent(final String eventName, final Context context, final String clientId) {
+        this(eventName);
+        setDefaults(context, clientId);
+    }
+
+    void setAuthority(final String authority) {
+        if (StringExtensions.isNullOrBlank(authority)) {
+            return;
+        }
+
+        setEvent(EventStrings.AUTHORITY_NAME, authority);
+        final URL authorityUrl = StringExtensions.getUrl(authority);
+        if (authorityUrl == null) {
+            return;
+        }
+
+        if (UrlExtensions.isADFSAuthority(authorityUrl)) {
+            setAuthorityType(EventStrings.AUTHORITY_TYPE_ADFS);
+        } else {
+            setAuthorityType(EventStrings.AUTHORITY_TYPE_AAD);
+        }
+    }
+
+    void setAuthorityType(final String authorityType) {
+        setEvent(EventStrings.AUTHORITY_TYPE, authorityType);
+    }
+
+    void setIsDeprecated(final Boolean isDeprecated) {
+        setEvent(EventStrings.API_DEPRECATED, isDeprecated.toString());
+    }
+
+    void setValidationStatus(final String validationStatus) {
+        setEvent(EventStrings.AUTHORITY_VALIDATION, validationStatus);
+    }
+
+    void setExtendedExpiresOnSetting(final Boolean extendedExpiresOnSetting) {
+        setEvent(EventStrings.EXTENDED_EXPIRES_ON_SETTING, extendedExpiresOnSetting.toString());
+    }
+
+    void setSilentRequestPromptBehavior(final String promptBehavior) {
+        setEvent(EventStrings.PROMPT_BEHAVIOR, promptBehavior);
+    }
+
+    void setAPIId(final String id) {
+        setEvent(EventStrings.API_ID, id);
+    }
+
+    String getEventName() {
+        return mEventName;
+    }
+
+    void setWasApiCallSuccessful(final Boolean isSuccess, final Exception exception) {
+        setEvent(EventStrings.WAS_SUCCESSFUL, isSuccess.toString());
+
+        if (exception != null) {
+            if (exception instanceof AuthenticationException) {
+                final AuthenticationException authException = (AuthenticationException) exception;
+                setEvent(EventStrings.API_ERROR_CODE, authException.getCode().toString());
+            }
```

Comment:
This would already get logged in the existing code.

---
In reply to: [90996260](https://github.com/AzureAD/azure-activedirectory-library-for-android/pull/736#discussion_r90996260) [](ancestors = 90996260)